lib/src/config_concatenation.cc

```diff
@@ -0,0 +1,246 @@
+#include <hocon/config_object.hpp>
+#include <hocon/path.hpp>
+#include <internal/config_concatenation.hpp>
+#include <internal/values/config_string.hpp>
+#include <internal/default_transformer.hpp>
+#include <internal/simple_config_list.hpp>
+#include <internal/simple_config_origin.hpp>
+#include <internal/config_exception.hpp>
+#include <internal/values/config_string.hpp>
+#include <internal/resolve_result.hpp>
+#include <internal/resolve_source.hpp>
+#include <internal/resolve_context.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    config_concatenation::config_concatenation(shared_origin origin, std::vector<shared_value> pieces) :
+            config_value(move(origin)), _pieces(move(pieces)) {
+        if (_pieces.size() < 2) {
+            throw config_exception("Created concatenation with less than 2 items");
+        }
+
+        bool had_unmergeable = false;
+        for (shared_value p : _pieces) {
+            if (dynamic_pointer_cast<const config_concatenation>(p)) {
+                throw config_exception("config_concatenation should never be nested");
+            }
+
+            if (dynamic_pointer_cast<const unmergeable>(p)) {
+                had_unmergeable = true;
+            }
+        }
+
+        if (!had_unmergeable) {
+            throw config_exception("Created concatenation without an unmergeable in it");
+        }
+    }
+
+    config_value_type config_concatenation::value_type() const {
+        throw not_resolved();
+    }
+
+    shared_ptr<vector<shared_value>> config_concatenation::unmerged_values() const {
+        return make_shared<vector<shared_value>>(vector<shared_value>{ shared_from_this() });
```

Comment:
Heh yeah, that came from http://stackoverflow.com/questions/28877488/vector-construction-when-wrapped-in-a-shared-pointer when I couldn't get the initializer to work by itself.